chronos/futures.nim

```diff
@@ -0,0 +1,369 @@
+  ## that this future belongs to, is a good habit as it helps with debugging.
+  let res = Future[T]()
+  setupFutureBase(res, getSrcLocation(fromProc))
+  res
+
+template completed*(
+    F: type Future, fromProc: static[string] = ""): Future[void] =
+  ## Create a new completed future
+  let res = Future[T]()
+  setupFutureBase(res, getSrcLocation(fromProc), FutureState.Completed)
+  res
+
+template completed*[T: not void](
+    F: type Future, valueParam: T, fromProc: static[string] = ""): Future[T] =
+  ## Create a new completed future
+  let res = Future[T](value: valueParam)
+  setupFutureBase(res, getSrcLocation(fromProc), FutureState.Completed)
+  res
+
+template failed*[T](
+    F: type Future[T], errorParam: ref CatchableError,
+    fromProc: static[string] = ""): Future[T] =
+  ## Create a new failed future
+  let res = Future[T](error: errorParam)
+  setupFutureBase(res, getSrcLocation(fromProc), FutureState.Failed)
+  res
+
+template cancelled*[T](
+    F: type Future[T], fromProc: static[string] = ""): Future[T] =
+  ## Create a new cancelled future
+  let res = Future[T](error: newCancelledError())
+  setupFutureBase(res, getSrcLocation(fromProc), FutureState.Cancelled)
+  res
+
+func finished*(future: FutureBase): bool {.inline.} =
+  ## Determines whether ``future`` has completed, i.e. ``future`` state changed
+  ## from state ``Pending`` to one of the states (``Finished``, ``Cancelled``,
+  ## ``Failed``).
+  future.state != FutureState.Pending
+
+func cancelled*(future: FutureBase): bool {.inline.} =
+  ## Determines whether ``future`` has cancelled.
+  future.state == FutureState.Cancelled
+
+func failed*(future: FutureBase): bool {.inline.} =
+  ## Determines whether ``future`` completed with an error.
+  future.state == FutureState.Failed
+
+func completed*(future: FutureBase): bool {.inline.} =
+  ## Determines whether ``future`` completed without an error.
+  future.state == FutureState.Completed
+
+when chronosStackTrace:
+  template getFilenameProcname(entry: StackTraceEntry): (string, string) =
+    when compiles(entry.filenameStr) and compiles(entry.procnameStr):
+      # We can't rely on "entry.filename" and "entry.procname" still being valid
+      # cstring pointers, because the "string.data" buffers they pointed to might
+      # be already garbage collected (this entry being a non-shallow copy,
+      # "entry.filename" no longer points to "entry.filenameStr.data", but to the
+      # buffer of the original object).
+      (entry.filenameStr, entry.procnameStr)
+    else:
+      ($entry.filename, $entry.procname)
+
+  proc getHint(entry: StackTraceEntry): string =
+    ## We try to provide some hints about stack trace entries that the user
+    ## may not be familiar with, in particular calls inside the stdlib.
+
+    let (filename, procname) = getFilenameProcname(entry)
+
+    if procname == "processPendingCallbacks":
+      if cmpIgnoreStyle(filename, "asyncdispatch.nim") == 0:
```

Comment:
asyncdispatch.nim - really?

Reply:
hey, this is straight from https://github.com/status-im/nim-chronos/blob/40143f8798e49e036179fbe066b0b32138ecfd85/chronos/asyncfutures2.nim#L420 - no idea what it does, but indeed, it looks out of date

Reply:
Yeah but you declared that you perform new `futures` so why should we keep this outdated/legacy stuff? Because i have also ideas how to improve different parts of code and so some of my comments in this review will be about improvements which you are not planned.

Reply:
removed